### PR TITLE
tasks tables in metadata storage are not cleared

### PR DESCRIPTION
**Pull request by seoeun25**
This PR mainly contains:

1. Remove the older tasks at metastore (druid_tasks) on tasklog cleanup schedule.
2. Not implement auditlog deletion since auditlogging is deprecated (#6368 )

### CODE REVIEW

File: core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java [hunk 1]
```diff
@@ -32,13 +32,14 @@
 {
   /**
    * Creates a new entry.
-   * 
-   * @param id entry id
-   * @param timestamp timestamp this entry was created
+   *
```
Comment (QiuMM) [inline core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java]: Why these changes, please don't modify code which is not related to your PR. The same for below.

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 2]
```diff
@@ -439,6 +439,26 @@ public Void withHandle(Handle handle)
     );
   }
 
+  @Override
+  public void removeTasksBefore(final DateTime createdTime)
```
[the remark this answers is not in the document]
Comment (seoeun25) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Have updated.

**Comment by seoeun25** [comment]
It marks @Deprecated to getSqlRemoveLogsOlderThan since tasklogs deprecated.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskStorage.java [hunk 3]
```diff
@@ -75,6 +75,13 @@
    */
   void removeLock(String taskid, TaskLock taskLock);
 
+  /**
+   * Remove the tasks created order than the given createdTime.
+   *
+   * @param timestamp timestamp to check the {@code created_date} of tasks
+   */
```
Comment (QiuMM) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskStorage.java]: I think the comment like below would be better:
```java
/**
   * Remove the tasks created order than the given timestamp.
   *
   * @param timestamp timestamp in milliseconds
   */
```

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 4]
```diff
@@ -439,6 +449,35 @@ public Void withHandle(Handle handle)
     );
   }
 
+  @Override
+  public void removeTasksOlderThan(final long timestamp)
+  {
+    connector.retryWithHandle(
+        (HandleCallback<Void>) handle -> {
+          DateTime createdTime = DateTimes.utc(timestamp);
```
Comment (QiuMM) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: I do not like this variable name, it's a time point not a created time. Maybe `DateTime dateTime = DateTimes.utc(timestamp);`

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 5]
```diff
@@ -439,6 +449,35 @@ public Void withHandle(Handle handle)
...
+          )
+                .bind("created_date", createdTime.toString())
+                .execute();
+          handle.createStatement(
+              StringUtils.format(
+                  "DELETE FROM %s WHERE created_date < :created_date AND active = false",
+                  entryTable
+              )
+          )
+                .bind("created_date", createdTime.toString())
```
Comment (QiuMM) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Same, change `created_date` to another name.

**Comment by QiuMM** [comment]
> It marks @deprecated to getSqlRemoveLogsOlderThan since tasklogs deprecated.

@seoeun25  Auditlog has been deprecated, not tasklogs.

File: core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java [hunk 6]
```diff
@@ -130,6 +130,13 @@ void insert(
    */
   void removeLock(long lockId);
 
+  /**
+   * Remove the tasks created order than the given createdTime.
```
Comment (surekhasaharan) [inline core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java]: nit: change "order" or "older". Or may be better way is "Remove the tasks with timestamp older than given timestamp

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskStorage.java (hunk 3 above)
Comment (surekhasaharan) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskStorage.java]: Agree, please change "order" to "older" here as well.

**Comment by gianm** [comment]
It looks like this is adding new behavior (deleting old tasks) to a pre-existing config (druid.indexer.logs.kill.durationToRetain). Is that right? What's the default for these? I'm just thinking about whether people will be surprised by the new behavior.

At any rate, I'll add a "release notes" label so we can call out what would change, if anything.

This patch should also include a doc update.

**Comment by seoeun25** [comment]
This commit contains:
1. change the variable name and comment.
2. update the doc to `druid.indexer.logs.kill.enabled`.

**Comment by QiuMM** [comment]
@jihoonson @seoeun25 oops, I had not checked it carefully, sorry.

**Comment by seoeun25** [comment]
@QiuMM @surekhasaharan @jihoonson @gianm Thanks for the review. Addressed comments and updated doc more detailed.

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 7]
```diff
@@ -439,6 +449,30 @@ public Void withHandle(Handle handle)
     );
   }
 
+  @Override
+  public void removeTasksOlderThan(final long timestamp)
+  {
+    DateTime dateTime = DateTimes.utc(timestamp);
+    log.info("Deleting tasks older than [%s] and inactive at metastore.", dateTime.toString());
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Sorry, I was not clear. I think it's better to remove `log` from this class (which shouldn't be part of this PR) because it's not very useful. Logging should give us detailed context about what's happening, but a simple `I will delete some tasks from metadata` message is not much useful I think. Rather, it looks good to me to print logs in the caller of this method, like in `TaskAutoCleaner`.

File: docs/content/configuration/index.md [hunk 8]
```diff
@@ -525,12 +525,12 @@ If you are running the indexing service in remote mode, the task logs must be st
...
 
-You can also configure the Overlord to automatically retain the task logs only for last x milliseconds by configuring following additional properties.
+You can also configure the Overlord to automatically retain the task logs in log directory and task-related metadata storage tables only for last x milliseconds by configuring following additional properties.
 Caution: Automatic log file deletion typically works based on log file modification timestamp on the backing store, so large clock skews between druid nodes and backing store nodes might result in un-intended behavior.  
 
 |Property|Description|Default|
 |--------|-----------|-------|
-|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. |false|
+|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. If set to true, overlord will submit tasks periodically based on `delay` specified. These kill tasks will delete task logs from log directory and tasks and tasklogs table in metadata storage except for the last `durationToRetain` period. |false|
```
Comment (surekhasaharan) [inline docs/content/configuration/index.md]: How about "If set to true, overlord will submit kill tasks periodically based on `delay` specified, which will delete task logs from the log directory as well as tasks and tasklogs table in metadata storage except for tasks created in the last `durationToRetain` period. "

File: docs/content/configuration/index.md [hunk 9]
```diff
@@ -525,12 +525,12 @@ If you are running the indexing service in remote mode, the task logs must be st
...
 |`druid.indexer.logs.kill.durationToRetain`| Required if kill is enabled. In milliseconds, task logs to be retained created in last x milliseconds. |None|
```
Comment (surekhasaharan) [inline docs/content/configuration/index.md]: If this behavior is also getting changed, please add doc here as well, something like "task logs and task-related metadata storage tables to be retained created in last x milliseconds"

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java (hunk 7 above)
Comment (seoeun25) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: I agree. I think it's better to remove `log`.

File: docs/content/configuration/index.md (hunk 8 above)
Comment (seoeun25) [reply docs/content/configuration/index.md]: Thanks for your comment !

**Comment by seoeun25** [comment]
@jihoonson @surekhasaharan Thanks for the review. remove redundant log and update doc more detailed.

File: core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java [hunk 10]
```diff
@@ -130,6 +130,13 @@ void insert(
    */
   void removeLock(long lockId);
 
+  /**
+   * Remove the tasks with timestamp older than given timestamp.
+   *
```
Comment (QiuMM) [inline core/src/main/java/org/apache/druid/metadata/MetadataStorageActionHandler.java]: `Remove the tasks created older than the given timestamp.` would be better.

File: docs/content/configuration/index.md [hunk 11]
```diff
@@ -525,13 +525,13 @@ If you are running the indexing service in remote mode, the task logs must be st
 |--------|-----------|-------|
 |`druid.indexer.logs.type`|Choices:noop, s3, azure, google, hdfs, file. Where to store task logs|file|
 
-You can also configure the Overlord to automatically retain the task logs only for last x milliseconds by configuring following additional properties.
+You can also configure the Overlord to automatically retain the task logs in log directory and task-related metadata storage tables only for last x milliseconds by configuring following additional properties.
 Caution: Automatic log file deletion typically works based on log file modification timestamp on the backing store, so large clock skews between druid nodes and backing store nodes might result in un-intended behavior.
```
Comment (QiuMM) [inline docs/content/configuration/index.md]: `...in log directory and entries in task-related metadata storage tables...` would be better.

File: docs/content/configuration/index.md [hunk 12]
```diff
@@ -525,13 +525,13 @@ If you are running the indexing service in remote mode, the task logs must be st
...
+You can also configure the Overlord to automatically retain the task logs in log directory and task-related metadata storage tables only for last x milliseconds by configuring following additional properties.
 Caution: Automatic log file deletion typically works based on log file modification timestamp on the backing store, so large clock skews between druid nodes and backing store nodes might result in un-intended behavior.  
 
 |Property|Description|Default|
 |--------|-----------|-------|
-|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. |false|
-|`druid.indexer.logs.kill.durationToRetain`| Required if kill is enabled. In milliseconds, task logs to be retained created in last x milliseconds. |None|
+|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. If set to true, overlord will submit kill tasks periodically based on `delay` specified, which will delete task logs from the log directory as well as tasks and tasklogs table in metadata storage except for tasks created in the last `durationToRetain` period. |false|
```
Comment (QiuMM) [inline docs/content/configuration/index.md]: I think
```
....based on `druid.indexer.logs.kill.delay` specified, which will delete task logs from the log directory as well as tasks and tasklogs table entries in metadata storage except for tasks created in the last `druid.indexer.logs.kill.durationToRetain` period. 
```
would be better.

File: docs/content/configuration/index.md [hunk 13]
```diff
@@ -525,13 +525,13 @@ If you are running the indexing service in remote mode, the task logs must be st
...
+|`druid.indexer.logs.kill.durationToRetain`| Required if kill is enabled. In milliseconds, task logs and task-related metadata storage tables to be retained created in last x milliseconds. |None|
```
Comment (QiuMM) [inline docs/content/configuration/index.md]: Same, ...task logs and entries in task-related....

**Comment by seoeun25** [comment]
@QiuMM Thanks for the comments! updated doc.
